src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderCursor.cs

```diff
@@ -750,15 +765,28 @@ private void ThreadProc()
                                     }
                                     line++;
 
-                                    // Filter out comments and empty strings.
+                                    // Filter out comments, empty strings and duplicate header string.
                                     if (text.Length >= 2)
                                     {
                                         // Don't use string.StartsWith("//") - it is too slow.
                                         if (text[0] == '/' && text[1] == '/')
+                                        {
+                                            _ch.Info("Ignore comments line {text} at line {line}.");
+                                            continue;
+                                        }
+
+                                        // Skip duplicate header string
+                                        if (_headerString == text)
+                                        {
+                                            _ch.Warning($"Ignore duplicate header string {text} at line {line}.");
```

Comment:
Although I guess this is the only way of actually solving the issue, I'm concerned about the performance implications of having to do a full string comparison for _every line_ we load. If the dataset is big (with many columns), would this performance implication would be significant?

Reply:
After thinking about it, I guess the implementation of comparing strings will actually be optimized to return "false" as soon as it finds a character that doesn't match... and in general even if the dataset has many columns, this comparison would return false very soon for the vast majority of the cases. So I guess this isn't a problem...

---
In reply to: [448687279](https://github.com/dotnet/machinelearning/pull/5275#discussion_r448687279) [](ancestors = 448687279)

Reply:
I agree with your assessment. But I would still like to see some benchmark test that this is true, especially on a large file.

---
In reply to: [448701405](https://github.com/dotnet/machinelearning/pull/5275#discussion_r448701405) [](ancestors = 448701405,448687279)

Reply:
Are we comparing the first string read in the header file with the other strings as they are read? What happens if there is another header row in the text file, but has different spacing between the commas or has quotes?

---
In reply to: [448759220](https://github.com/dotnet/machinelearning/pull/5275#discussion_r448759220) [](ancestors = 448759220,448701405,448687279)

Reply:
I'm not sure that also checking for variations of the header (spacing, quotes, capitalization, culture variation, etc ...) would be a good idea, as it would necessarily increase the time spent while reading _every line_ .
Besides, I'm not sure how much coverage we should give to fix #5051 ? I'd think it isn't a very common problem anyway, and so simply checking for the exact same string as found on the header of the file would be enough.

---
In reply to: [448759887](https://github.com/dotnet/machinelearning/pull/5275#discussion_r448759887) [](ancestors = 448759887,448759220,448701405,448687279)